resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+modified to account for explicit nullability and the accompanying type system
+changes (including the legacy types).  In the interest of backwards
+compatibility, it does not try to fix the various issues with the existing
+algorithm.  
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+## Helper predicates
+
+The **TOP** predicate is true for any type which is in the equivalence class of
+top types.
+
+- **TOP**(`T?`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`T*`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`dynamic`) is true
+- **TOP**(`void`) is true
+- **TOP**(`FutureOr<T>`) is **TOP**(T)
+- **TOP**(T) is false otherwise
+
+The **OBJECT** predicate is true for any type which is in the equivalence class
+of `Object`.
+
+
+- **OBJECT**(`Object`) is true
+- **OBJECT**(`dynamic`) is true
+- **OBJECT**(`void`) is true
+- **OBJECT**(`FutureOr<T>`) is **OBJECT**(T)
+- **OBJECT**(T) is false otherwise
+
+The **BOTTOM** predicate is true for things in the equivalence class of `Never`.
+
+- **BOTTOM**(`Never`) is true
+- **BOTTOM**(`X&T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`X extends T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`T`) is false otherwise
+
+The **NULL** predicate is true for things in the equivalence class of `Null`
+
+- **NULL**(`Null`) is true
+- **NULL**(`T?`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T*`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on top and `Object` types.
+
+- **MORETOP**(`void`, `T`) = true
+- **MORETOP**(`T`, `void`) = false
+- **MORETOP**(`dynamic`, `T`) = true
+- **MORETOP**(`T`, `dynamic`) = false
+- **MORETOP**(`Object`, `T`) = true
+- **MORETOP**(`T`, `Object`) = false
+- **MORETOP**(`T?`, `S?`) = **MORETOP**(`T`, `S`)
+- **MORETOP**(`T?`, `S`) = true
+- **MORETOP**(`T`, `S?`) = false
+- **MORETOP**(`T*`, `S*`) = **MORETOP**(`T`, `S`)
+- **MORETOP**(`T*`, `S`) = true
+- **MORETOP**(`T`, `S*`) = false
+- **MORETOP**(`FutureOr<T>`, `FutureOr<S>`) = **MORETOP**(T, S)
+
+The **MOREBOTTOM** predicate defines a total order on bottom and `Null` types.
+
+- **MOREBOTTOM**(`Never`, `T`) = true
+- **MOREBOTTOM**(`T`, `Never`) = false
+- **MOREBOTTOM**(`Null`, `T`) = true
+- **MOREBOTTOM**(`T`, `Null`) = false
+- **MOREBOTTOM**(`T?`, `S?`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`T?`, `S`) = true
+- **MOREBOTTOM**(`T`, `S?`) = false
+- **MOREBOTTOM**(`T*`, `S*`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`T*`, `S`) = true
+- **MOREBOTTOM**(`T`, `S*`) = false
+- **MOREBOTTOM**(`X&T`, `X&S`) = **MOREBOTTOM**(`T`, `S`)
+- **MOREBOTTOM**(`X&T`, `S`) = true
+- **MOREBOTTOM**(`S`, `X&T`) = false
+- **MOREBOTTOM**(`X extends T`, `X extends S`) = **MOREBOTTOM**(`T`, `S`)
```

Comment:
We would need a rule where the two type variables differ (otherwise they'd presumably have the same bound as well).

Similarly, if we can't match different type variables in the `X&T` case then some top types will presumably just be unrelated (and then we don't have a total order).

Reply:
This is a bit ugly, unfortunately.  It means that we can't be completely symmetric and still respect alpha-variance.  That is, what do I do with MOREBOTTOM(`X extends Never`, `Y extends Never`)?  The only difference that I could order based on is the name of the type variable, which I really don't want to do .  Or I guess something silly like which one was declared first but.... yuck.

This is another point in favor of normalization - they both nicely normalize to `Never`.